src/bufr/IodaEncoder/IodaDescription.h

```diff
@@ -41,8 +42,16 @@ namespace Ingester
         int compressionLevel;  // Optional
     };
 
+    struct GlobalDescription
```

Comment:
If you feel comfortable with c++ templates consider making this a template so that the value can be of the type specified in the config. That way you could get rid of the "type" property altogether and it would save you work later converting strings to number types etc... Be aware if you do this you will need to make a GlobalDescriptionBase (non-template) to inherit from so you can put them in lists as shared_ptrs (not the direct values or you might end up slicing the object later). ex:

```cpp
 // IodaDescription.h

 struct GlobalDescriptionBase
 {
    std::string name;
 };

 template<typename T>
 struct GlobalDescription : GlobalDescriptionBase
 {
    T value;
 };

 //....
 std::vector<std::shared_ptr<GlobalDescriptionBase>> globals_;


 // IodaDescription.cpp

 if (conf.getString(ConfKeys::Global::Name) == ConfKeys::Global::StringType)
 {
     auto global = std::make_shared<GlobalDescription<std::string>>();
     global->name = // get the name

     // get the value string from the YAML.
     // validate it and convert it to the correct type (string is trivial)
     global->value = my_value 

     globals_.push_back(global)
 }
```

If all this sounds really overwhelming you can keep it the way it is.

Reply:
Thank you Ron! I have been working on implementing this but didn't quite know how to use the C++ templating. What you have here really helps and hopefully I can get this done correctly.

Reply:
No problem, let me know if you run into any issues.

Reply:
@rmclaren I ran into a problem when defining vector 
```cpp
std::vector<std::shared_ptr<GlobalDescriptionBase>> globals_;
```
The compiler asks me to include arguments for GlobalDescriptionBase and I wrote 
```cpp
std::vector<std::shared_ptr<GlobalDescriptionBase<std::string>>> globals_;
```
to make it compile. But after this I found it converts all values in struct ```GlobalDescription``` into ```cpp std::string```. 

The latest commit I implemented it in a different way which is closer to the original code, but the issue remains. When I try to print the type of the value after getFloat it gives me this: 
```
NSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEE
```
I suppose it is string. 

Currently the string option completely works but the float option doesn't work.

Reply:
Could you please push your code. It sounds like you made your GlobalDescriptionBase a templated struct... you do not want this. Only its bubclass GlobalDescription should be templated.

Reply:
What you have in the repo right now won't work. Please look at the coding example I provided.

Reply:
Don't forget to #include <memory> at the top of the file.

Reply:
I was trying and error and pushed a wrong version I think. I'll go back to the previous version which is following your example.

Reply:
As a side note: You actually don't need to include type_traits in IodaDescription.h, that was kinda a typo on my part.

Reply:
I just rewinded it to the previous commit and added the code try to handle floats. The two problems I have are as of below: 
I can't compile the code when writing the if statement in ``` IodaEncoder.cpp``` as: 
```cpp
if (std::is_same<std::string, global->value>::value)
{
}
else if (std::is_same<float, global->value>::value)
{
}
```
The compiler complains on ```global``` not being a typename. So I passed the ```global->type``` to get around it. 

Then if I implemented the above work around, I can't compile the code when I write the float attributes: 
```cpp
attr.write<float>({global->value});
```
The compiler will complain about the argument type. I printed its value and type above and the result is as below: 
```
e
NSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEE
```
I think this means it is still a string which goes back to the problem I talked about above.

Reply:
So one problem is line 77 in IodaDescription.h: std::vector<std::shared_ptr<GlobalDescription<std::string>>> globals_;

You actually want to specify GlobalDescriptionBase not GlobalDescription<std::string>.

I'll go ahead and make some modifications to your code and check them into this repository to help speed things up a bit (will have to be later, you'll have my changes by Monday). If you want we can have a call and I can explain what is going on.

Basically a template struct (or class), is like a prototype for that class.  So when you specify a template class like above GlobalDescription<std::string> the compiler creates that class for you at compile time (there is literally now a new kind of class GlobalDescription__String (we could call it)). Templates are literally functions and classes which tell the compiler how to automatically generates new functions and classes on the fly :) So its almost like a meta programming language, literally code that the compiler uses to create more code. As such your mindset has to be a bit different when thinking about these things. For instance you can't just stick a template in a like this vector<GlobalDescription>, because that thing GlobalDescription is literally not defined (its a template for a class, not a class). GlobalDescritpion<String> of course is actually a class, but if we stick that in a vector (like above) then that vector can only hold the version of GlobalDescription the compiler made that can hold strings :) So what to do... well we can use a concept called polymorphism to make a vector that can hold any kind of GlobalDescription. When one object inherits from another object it **is a** type of that object. So the way we've arranged things above GlobalDescription<String>, GlobalDescription<Float>, and so on are all types of GlobalDescriptionBase, even though the compiler generated them on the fly. Cool, so we can have the compiler make a vector<shared_ptr<GlobalDescriptionBase>> for us, so we can hold any kind of global description. Anyways, I could go one. We can talk more on Monday. Take care!

Reply:
Thank you so much Ron!! That's actually where I suspect the problem is but don't know how to correct it. I have spent quite some time on it and It looks like the compiler doesn't like me to pass explicit argument when defining that vector, nor it allows me to pass no argument as well. If you can quickly show me how to work around I think will help a lot.  Thanks again!

Reply:
@rmclaren  I have finished the implementation of int type as well as the floatVector and intVector. However, when I test it to write a vector type of attribute it will report the following error: 
```
terminate called after throwing an instance of 'std::logic_error'
  what():  Bad assertion / unsupported fundamental type at the frontend side of the ioda type system.
Aborted
```
I believe this has something to do with the Ioda frontend. I printed the vector type and vector values and they look good to me but just can't use ```ioda::Attribute attr = group.atts.create<T>(name, {vectorLength});``` to create the attribute. Maybe @srherbener can help here? 
To reproduce the error, just add a global attribute in the global section as: 
```yaml
      globals:
        - name: "sensorCentralWavelength@MetaData"
          type: floatVector
          value: [101.5, 102.5, 103.5]
```